### PR TITLE
Vendor `rubocop-ast`

### PR DESCRIPTION
**Pull request by SeekingMeaning**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

Currently, `Utils::AST` is calling `Homebrew.install_bundler_gems!` to get access to `rubocop-ast`. Also, `formula_component_before_target?` requires `rubocops/components_order` to get access to `COMPONENT_PRECEDENCE_LIST`. This indirectly creates a requirement for `rubocop`, which also requires `Homebrew.install_bundler_gems!`

This is not ideal since it slows down homebrew-core publish jobs.

This PR adds `rubocop-ast` and its dependencies (`parser` and `ast`) as vendored gems and adds `ast_constants.rb` so that `Homebrew.install_bundler_gems!` is no longer necessary

### CODE REVIEW

**Review by dawidd6 — APPROVED** [review]

File: Library/Homebrew/Gemfile [hunk 1]
```diff
@@ -22,10 +22,13 @@ gem "tapioca", require: false
 
 # vendored gems
 gem "activesupport"
+gem "ast"
 gem "concurrent-ruby"
 gem "mechanize"
+gem "parser"
```
Comment (reitermarkus) [inline Library/Homebrew/Gemfile:28]:
```suggestion
```

File: Library/Homebrew/Gemfile [hunk 2]
```diff
@@ -22,10 +22,13 @@ gem "tapioca", require: false
 
 # vendored gems
 gem "activesupport"
+gem "ast"
```
Comment (reitermarkus) [inline Library/Homebrew/Gemfile:25]:
```suggestion
```

**Review by reitermarkus — COMMENTED** [review]
Given these are already dependencies of `rupbocop-ast`, no need to specify them explicitly.

**Comment by reitermarkus** [comment]
> This is not ideal since it slows down homebrew-core publish jobs.

Can we not cache the gems in there? Adding 133k lines to this repo just to speed up publish jobs feels kind of overkill.

**Comment by SeekingMeaning** [comment]
> > This is not ideal since it slows down homebrew-core publish jobs.
> 
> Can we not cache the gems in there? Adding 133k lines to this repo just to speed up publish jobs feels kind of overkill.

Ping @dawidd6

**Comment by dawidd6** [comment]
> Can we not cache the gems in there? Adding 133k lines to this repo just to speed up publish jobs feels kind of overkill.

Yes we can actually, that was my first idea. Dunno which solution would be better tho.

**Comment by reitermarkus** [comment]
> Dunno which solution would be better tho.

I'd say the one which doesn't add 133k lines of code to every Homebrew installation, especially since this is only needed for `dev-cmd`s and GitHub Actions.

**Comment by SeekingMeaning** [comment]
Related to Sorbet method signatures: Several methods in `utils/ast.rb` have a parameter of type `RuboCop::AST::Node` but we can't reference that class in `sig` blocks since `install_bundler_gems` needs to be called first. Any advice @reitermarkus?

**Comment by reitermarkus** [comment]
When are we requiring `utils/ast`? Can we require `rubocop/ast` before that point?

**Comment by SeekingMeaning** [comment]
Currently only `dev-cmd/bump-revision` and `dev-cmd/bottle` are using `utils/ast` but I plan to use it in more commands (e.g. `bump-formula-pr`)

**Comment by reitermarkus** [comment]
But where is the `require "utils/ast"` line? It should work if we first call `install_bundler_gems` and only then require `utils/ast`.

**Comment by SeekingMeaning** [comment]
- Here: https://github.com/Homebrew/brew/blob/acf0ecc/Library/Homebrew/dev-cmd/bump-revision.rb#L6
- and here: https://github.com/Homebrew/brew/blob/e5eb6a2/Library/Homebrew/dev-cmd/bottle.rb#L12

But we could move it into the `bump_revision` and `merge` methods:

- https://github.com/Homebrew/brew/blob/acf0ecc/Library/Homebrew/dev-cmd/bump-revision.rb#L52
- https://github.com/Homebrew/brew/blob/e5eb6a2/Library/Homebrew/dev-cmd/bottle.rb#L495

It's not ideal but we'll have to call `install_bundler_gems` wherever `utils/ast` is required

**Comment by reitermarkus** [comment]
> But we could move it into the `bump_revision` and `merge` methods:

That should work. `ast_spec.rb` should be fine since `brew tests` installs the gems beforehand.

**Comment by MikeMcQuaid** [comment]
> I'd say the one which doesn't add 133k lines of code to every Homebrew installation, especially since this is only needed for `dev-cmd`s and GitHub Actions.

Agreed. As a general rule we only want to vendor gems needed for `cmd`s and not `dev-cmd`s (most of which already install all bundler gems).

**Comment by SeekingMeaning** [comment]
Superseded by #10241

**Pull request closed without merging**